caledonia/src/bounded.rs

```diff
@@ -0,0 +1,451 @@
+        }
+        for &s in set.iter() {
+            // TODO: add H1(s),
+            // let (hs, index_s) = Proof::h0(setup, s);
+            bins[Proof::h0(setup, v, s)].push(s);
+        }
+        // Initialising max number of steps
+        let mut limit = setup.b;
+        // Attempting to generate a proof for $d$ rounds
+        for t in 1..(setup.d + 1) {
+            if limit == 0 {
+                return (0, None);
+            }
+            limit = limit - 1;
+            let round = Round::new(v, t, setup.n_p);
+            let res = Proof::dfs(setup, &bins, &round, &mut limit);
+            if res.is_some() {
+                return (limit, res);
+            }
+        }
+        // If no proof found, return None
+        return (limit, None);
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
+    /// Returns an empty proof if no suitable candidate is found.
+    pub fn prove(setup: &Setup, set: &Vec<[u8; 32]>) -> Self {
+        // Initialising our n_p bins with all s ∈ Sp
+        for v in 0..setup.r {
+            if let (_, Some(proof)) = Proof::prove_index(setup, set, v) {
+                return proof;
+            }
+        }
+        // If no proof found, return empty proof
+        return Proof::empty();
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
+    /// Returns an empty proof if no suitable candidate is found.
+    pub fn bench(setup: &Setup, set: &Vec<[u8; 32]>) -> (usize, Self) {
+        // Initialising our n_p bins with all s ∈ Sp
+        let mut nb_calls = 0;
+        for v in 0..setup.r {
+            let (steps, opt) = Proof::prove_index(setup, set, v);
+            // Adding nb of steps done in prove_index
+            nb_calls += setup.b - steps;
+            if let Some(proof) = opt {
+                return (nb_calls, proof);
+            }
+        }
+        // If no proof found, return empty proof
+        return (nb_calls, Proof::empty());
+    }
+
+    /// Alba's verification algorithm, follows proving algorithm by running the
+    /// same depth-first search algorithm.
+    pub fn verify(setup: &Setup, proof: Proof) -> bool {
+        if proof.d == 0 || proof.d > setup.d || proof.r > setup.r || proof.items.len() != setup.u {
+            return false;
+        }
+        let r0 = Round::new(proof.r, proof.d, setup.n_p);
+        let (b, round) = proof.items.iter().fold((true, r0), |(b, r), &s| {
+            (
+                b && r.h_usize == Proof::h0(setup, proof.r, s),
+                Round::update(&r, s),
+            )
+        });
+        return b && Proof::h2(setup, &round);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use rand_chacha::ChaCha20Rng;
+    use rand_core::{RngCore, SeedableRng};
+
+    #[test]
+    fn test_verify() {
+        let mut rng = ChaCha20Rng::from_seed(Default::default());
+        let nb_tests = 1_000;
+        for _t in 0..nb_tests {
```

Comment:
Perhaps using a property based testing library would make the test simpler?

Reply:
Definitely! There is an issue for this. For the time being, it's good to have this test however.